### DISCUSSION TITLE
[version 7.3.4] CheckBox undefined value in form data

### DISCUSSION THREAD

**Discussion by fabrusso** [react-hook-form/react-hook-form#4979] [Q&A] [answered]
**Describe the bug**
I'm trying to connect a CheckBox input to an useForm with {...register("checkbox")} as name.
Currently the checkbox is always undefined, also if the checked property is passed as true.

**To Reproduce**

1. Create an useForm form
2. Add a CheckBox
3. console.log the form data to console
4. Checkbox value is undefined

Sandbox Demo:
https://codesandbox.io/s/magical-sea-9t8vf?file=/src/App.js

**Expected behavior**
Show the value of checkbox as `true` or `false`, or checked and not checked

**Devices**
Trying this on 86.0b6 (64 bit) Mozilla for Developers
Package version: 7.3.4

**Comment by bluebill1049** [comment]
plz use `useController` or `Controller` for controlled input.

https://react-hook-form.com/api/usecontroller
https://react-hook-form.com/api/usecontroller/controller

**Comment by fabrusso** [comment]
> plz use `useController` or `Controller` for controlled input.
> 
> https://react-hook-form.com/api/usecontroller
> https://react-hook-form.com/api/usecontroller/controller

I modified the Demo with the `<Controller />` component:

https://codesandbox.io/s/magical-sea-9t8vf?file=/src/App.js

```
<Controller
  control={control}
  name="checkbox"
  render={({ field: { onChange, value, name, ref } }) => (
    <Input name={name} value={value} ref={ref} onChange={onChange} />
   )}
 />
```

It's still returning `undefined`

**Comment by bluebill1049** [comment]
where is the onChagne prop in your custom component? 

<img width="554" alt="Screen Shot 2021-04-27 at 10 05 08 pm" src="https://user-images.githubusercontent.com/10513364/116238400-a2a30580-a7a4-11eb-8501-c67ffc401ae3.png">

**Comment by bluebill1049** [comment]
have a read at this section: https://react-hook-form.com/advanced-usage#TransformandParse I think you will understand better.

**Comment by bluebill1049** [comment, marked as answer]
here is a working version: https://codesandbox.io/s/lingering-architecture-86459 

what's passing into the `onChange` is what sent to hook form.

**Reply by fabrusso** [reply]
Oh... now I see.

I will try this also in my original project to better understand.
I got to past the `onChange` to the Controlled component from the `onChange` into my component.

**Reply by fabrusso** [reply]
> here is a working version: https://codesandbox.io/s/lingering-architecture-86459
> 
> what's passing into the `onChange` is what sent to hook form.

Is It possible not to have `undefined` initially?

Also If my initialState is `false` the form data has `undefined` as value

I see I can use `defaultValue` on the Controller, is that the only way?

**Reply by bluebill1049** [reply]
we don't support defaultValue as `undefined`, as native input really shouldn't have value `undefined` and we need the undefined check to fall back to the global defaultValues at useForm, you can perhaps use `null`

**Reply by FDiskas** [reply]
Find the difference - one works another don't :D
https://codesandbox.io/s/magical-sea-9t8vf?file=/src/App.js
vs 
https://codesandbox.io/s/lingering-architecture-86459?file=/src/App.js

**Reply by FDiskas** [reply]
the simplified version https://codesandbox.io/s/peaceful-cray-x33ny?file=/src/Input.js

**Reply by sonphung205** [reply]
> Find the difference - one works another don't :D
> https://codesandbox.io/s/magical-sea-9t8vf?file=/src/App.js
> vs
> https://codesandbox.io/s/lingering-architecture-86459?file=/src/App.js

OMG they have no difference and in same version!
